Simplify and shrink docker setup

erijo opened the pull request:
### What does this PR do?

* Remove the requirement on docker-compose for running tests.
* Shrink the docker image with ~380 MB (no apt recommendations + some other minor changes).

### What issues does this PR fix or reference?

None.

### Previous Behavior

A bigger than necessary docker image.

### New Behavior

The image is now 251 MB instead of 637 MB.

### Have [tests][1] been written for this change?

No, but the existing tests pass on the new image.

### Have these commits been [signed with GnuPG][2]?

Yes

---

Please review [yadm's Contributing Guide][3] for best practices.

[1]: https://github.com/TheLocehiliosan/yadm/blob/master/.github/CONTRIBUTING.md#test-conventions
[2]: https://help.github.com/en/articles/signing-commits
[3]: https://github.com/TheLocehiliosan/yadm/blob/master/.github/CONTRIBUTING.md

TheLocehiliosan commented:
I think there are a lot of good changes here, but I think the docker image could be shrunk without all of the `&&`, `&&`, `&&`... The joining of all of these operations to achieve less layers makes the Dockerfile harder to read (at least for me).

Could you keep the easier to read multiple layers, along with the improvements like xz-utils, `--no-install-recommends`, and moving the install of `locales`, but then we can shrink the image by.

```
FROM ubuntu:18.04 as build
...
...
FROM scratch
COPY --from=build / /
WORKDIR /yadm
CMD make test
```

I think this will result in an even smaller image. How does that sound?

erijo commented:
Sure thing. I started merging the layers and even if it saved some space it was skipping recommends that paid off the most. I'll update the pr later.

TheLocehiliosan commented:
I'm not sure without testing, but it may also be necessary to move the `ENV` that sets locale info and `MAINTAINER` for those elements to get into the final image. `MAINTAINER` can just be moved, but I'm not sure if those locale settings are necessary during the build (I think they are not, so it can probably just be moved).

`WORKDIR` and `CMD` will for sure need to be after the "from scratch copy", which probably means `/yadm` dir needs to be created before the fallback stub `/yadm/Makefile` is created.

erijo commented:
I've restored some, but not all, of the RUN statements. Commands that belong together are still done in a single RUN. But I switched to using ADD to download files and could then remove curl from the image. I also switched to vim-tiny instead of vim and all together the image is now even smaller than it was in the first attempt.

I tried the `COPY --from` trick and it only saved 2 MB so I skipped that (as it seemed a bit magic).

TheLocehiliosan commented:
Sounds good. I might want to test that out later, I sometimes use my own dotfiles to do some testing, and that has a bootstrap that installs vim plugins using vim-plug. Not sure if vim-tiny would have any issue with that.

erijo commented:
I added two more changes now: one to include old versions of yadm to help test #277 and one to set up git to avoid having to do that manually each time you want to commit something. You can skip them if you don't think they are needed.
